lib/semantics/tools.cc

```diff
@@ -794,4 +761,139 @@ static Symbol &InstantiateSymbol(
   return result;
 }
 
+enum class ComponentKind { Direct, Ultimate, Potential };
+
+template<ComponentKind componentKind> class ComponentVisitorImplementation {
+public:
+  ComponentVisitorImplementation(std::function<bool(const Symbol &)> &predicate)
+    : predicate_{predicate} {}
+  SymbolVector Visit(const DerivedTypeSpec &derived) {
+    TraverseDerivedType(derived);
+    return std::move(componentStack_);
+  }
+
+private:
+  const Symbol *TraverseDerivedType(const DerivedTypeSpec &derived) {
+    const Symbol &derivedTypeSymbol{derived.typeSymbol()};
+    // Avoid infinite loop if the type is already part of the types
+    // being visited. It is possible to have "loops in type" because
+    // C744 does not forbid to use not yet declared type for
+    // ALLOCATABLE or POINTER components.
+    // When looking for potential components, the search could fall
+    // in an infinite loop. Avoid these in other search for safety.
+    for (const Symbol *typeUnderVisit : typeStack_) {
+      if (typeUnderVisit == &derivedTypeSymbol) {
+        return nullptr;
+      }
+    }
+    typeStack_.push_back(&derivedTypeSymbol);
+    const Scope *scope{derivedTypeSymbol.scope()};
+    CHECK(scope);  // derived type symbol must have a scope
+    const Symbol *result{nullptr};
+    // Note: parent subcomponents are visited first, then the parent component
+    // (if any), then other components by order of declaration
+    SymbolVector orederedComponents{
+        derivedTypeSymbol.get<DerivedTypeDetails>().OrderComponents(*scope)};
+    for (const Symbol *component : orederedComponents) {
+      if (component->has<ProcEntityDetails>()) {
+        result = VisitProcedureComponent(*component);
+      } else if (component->has<ObjectEntityDetails>()) {
+        result = VisitDataComponent(*component);
+      }
+      if (result) {
+        return result;
```

Comment:
It's surprising to push the type but not always pop it. Maybe add a comment about why the pop is not necessary (assuming it's not).

Reply:
> An iterator for visiting components in order would also work as an alternative design pattern for your visitation framework, and could be used to replace `DerivedTypeDetails::OrderComponents()`, which today returns a vector of components.

That is an interesting idea, thanks.
So far, the current visiting framework is actually using `DerivedTypeDetails::OrderComponents()` to traverse derived type sub-components. But, if I transform the ultimate/direct/potential visitor into iterators, they would probably live inside the `DerivedTypeDetails` or at least be `friend` and would not require `OrderComponents`.

I guess I can transform the visitors into iterators by transforming the `componentStack_` into some kind of stack of iterators over `DerivedTypeDetails::componentNames_` and by removing the predicate testing part (and probably a bit more).
I will explore the idea.

Reply:
Oops, the reply above did not belong here. The pop was not necessary, but yes a comment would have been nice.